Remove `kw::Empty` uses from `hir::Lifetime::ident`

nnethercote opened the pull request:
`hir::Lifetime::ident` is sometimes set to `kw::Empty` and it's really confusing. This PR stops that. Helps with #137978.

r? @lcnr

nnethercote committed:
Don't use `kw::Empty` in `hir::Lifetime::ident`.

`hir::Lifetime::ident` currently sometimes uses `kw::Empty` for elided
lifetimes and sometimes uses `kw::UnderscoreLifetime`, and the
distinction is used when creating some error suggestions, e.g. in
`Lifetime::suggestion` and `ImplicitLifetimeFinder::visit_ty`. I found
this *really* confusing, and it took me a while to understand what was
going on.

This commit replaces all uses of `kw::Empty` in `hir::Lifetime::ident`
with `kw::UnderscoreLifetime`. It adds a new field
`hir::Lifetime::is_path_anon` that mostly replaces the old
empty/underscore distinction and makes things much clearer.

Some other notable changes:

- Adds a big comment to `Lifetime` talking about permissable field
  values.

- Adds some assertions in `new_named_lifetime` about what ident values
  are permissible for the different `LifetimeRes` values.

- Adds a `Lifetime::new` constructor that does some checking to make
  sure the `is_elided` and `is_anonymous` states are valid.

- `add_static_impl_trait_suggestion` now looks at `Lifetime::res`
  instead of the ident when creating the suggestion. This is the one
  case where `is_path_anon` doesn't replace the old empty/underscore
  distinction.

- A couple of minor pretty-printing improvements.

nnethercote committed:
Remove `LifetimeSuggestionPosition` and `Lifetime::suggestion_position`.

They both are only used in `Lifetime::suggestion`. This commit inlines
and removes them.

nnethercote committed:
Remove `ImplicitObjectLifetimeDefault` case from `suggestion`.

It has no effect on anything in the test suite.

This means it can also be rewritten as a neater pairwise `match`.

nnethercote committed:
Add a HIR pretty-printing test focused on lifetimes.

HIR printing currently gets very little testing. This increases coverage
a bit, with a focus on lifetimes.

There are some FIXME comments for cases that are printed in a dubious
fashion. This PR won't address those; the point of adding this test is
to ensure that the subsequent commits don't hurt pretty-printing.

nnethercote commented:
Best reviewed one commit at a time.

lcnr commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -1777,7 +1778,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+        self.new_named_lifetime(l.id, l.id, l.ident, IsPathAnon::No)
+    }
+
+    fn lower_lifetime_is_path_anon(&mut self, id: NodeId, span: Span) -> &'hir hir::Lifetime {
```
```suggestion
    fn lower_lifetime_path_anon(&mut self, id: NodeId, span: Span) -> &'hir hir::Lifetime {
```
or `lower_lifetime_in_path_anon` or sth 😅 i'd expect a function with "_is_X" to have an argument to set X

nnethercote replied:
I'll do `lower_lifetime_anon_in_path` and `IsAnonInPath` for the typename.

lcnr commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -1786,28 +1791,38 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+                hir::LifetimeName::Infer
+            }
+            LifetimeRes::Static { .. } => {
+                debug_assert!(matches!(ident.name, kw::StaticLifetime | kw::UnderscoreLifetime));
```
when does this happen to be an underscore lifetime?

nnethercote replied:
When the `'static` is elided, e.g. for both `C` and `D` here:
```
struct Name<'a>(&'a str);
const C: Name = Name("name"); // IsAnonInPath::Yes
static D: &str = "";          // IsAnonInPath::No
```
I didn't even realize `'static` could be elided like this!

I will add a comment about this.

lcnr commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -35,20 +35,27 @@ use crate::def_id::{DefId, LocalDefIdMap};
+
+    /// Is the lifetime anonymous and in a path? Used only for error
+    /// suggestions. See `Lifetime::suggestion` for example use.
+    pub is_path_anon: IsPathAnon,
```
`IsAnonInPath`? `IsPathAnon` feels very ambiguous to me

nnethercote replied:
Ok, `IsAnonInPath` works.

lcnr commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2391,11 +2406,12 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+        let r = hir::Lifetime::new(
+            self.next_id(),
+            Ident::new(kw::UnderscoreLifetime, self.lower_span(span)),
+            hir::LifetimeName::ImplicitObjectLifetimeDefault,
+            IsPathAnon::No,
```
would it be easier to have a `LifetimeName::Anon|ElidedInPath`? and not have two enums inside of `hir::Lifetime`?

nnethercote replied:
I originally wanted to combine the two enums in exactly this fashion, but it turns out that `IsAnonInPath::Yes` can occur with four of the five `LifetimeName` variants, viz:
```rust
#![allow(unused)]

//---------------------------------------------------------------------------
struct Name<'a>(&'a str);
const C: Name = Name("name"); // LifetimeRes::Static, name='_, IsAnonInPath::Yes
static D: &str = "";          // LifetimeRes::Static, name='_, IsAnonInPath::No

//---------------------------------------------------------------------------
#[repr(C)]
struct Safe<'a>(&'a u32);
extern "C" {
    fn foo(safe: Safe);     // LifetimeRes::Param, name='_, IsAnonInPath::Yes
}

//---------------------------------------------------------------------------
struct St<'a> { f1: &'a u32 }

fn f() {
    let st = St { f1: &0u32 };  // LifetimeRes::Infer, name='_, IsAnonInPath::Yes
}

//---------------------------------------------------------------------------
// commented out because it causes an error
//struct Baz<'a>(&'a str);    // LifetimeRes::Error, name='_, IsAnonInPath::Yes
//struct Baz2(Baz);

fn main() {}
```
Once again, lifetimes can be omitted from paths in more places than I realized.

If you wanted to combine `ident` *and* `res` *and* `is_anon_in_path` you'd have an enum like this:
```rust
enum Combined {
    ParamUserNamed(Ident),         // user-specified name, IsAnonInPath::No
    ParamAnon(Span, IsAnonInPath), // name='_

    Infer(Span, IsAnonInPath),     // name='_

    StaticNamed(Span),             // name='static, IsAnonInPath::No
    StaticAnon(IsAnonInPath),      // name='_

    Error(Ident),                  // user-specified name, IsAnonInPath::No
    ErrorAnon(Span, IsAnonInPath), // name='_

    ImplicitObjectLifetimeDefault, // name='_, IsAnonInPath::No
}
```
I *think* that captures all the possibilities and excludes all invalid combinations. I'm generally very much a fan of "don't let your types express invalid values" but this feels like it's taking it too far?

nnethercote commented:
I updated, making these changes:
- Renamed `IsPathAnon` to `IsAnonInPath`.
- Added a big comment on `Lifetime`.
- Added an assertion to `new_named_lifetime`: if `IsAnonInPath::Yes`, the lifetime name must be `'_`.
- Expanded pretty-printing test coverage. The main commit now makes some slight improvements to pretty-printing.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14099247460/job/39492331784) [(plain)](https://github.com/rust-lang/rust/commit/4f1b408d87b8c1c3d92d002268a467454994e42f/checks/39492331784/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.6s done
#19 DONE 33.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.227
[RUSTC-TIMING] test test:false 9.276
    Finished `release` profile [optimized] target(s) in 1m 21s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 81.575
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.828
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 3.017
    Finished `release` profile [optimized] target(s) in 4.14s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.198
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.187
---
---- compiler/rustc_hir/src/hir.rs - hir::Lifetime (line 46) stdout ----
error: invalid label name `'_`
  --> compiler/rustc_hir/src/hir.rs:69:59
   |
25 | ...                   name='_, IsAnonInPath::No
   |                            ^^

error: expected `while`, `for`, `loop` or `{` after a label
  --> compiler/rustc_hir/src/hir.rs:69:61
   |
25 | ...                   name='_, IsAnonInPath::No
   |                              ^ expected `while`, `for`, `loop` or `{` after a label
   |
help: add `'` to close the char literal
   |
25 |                                                      name='_', IsAnonInPath::No
   |                                                             +

error: expected one of `.`, `:`, `;`, `?`, `for`, `loop`, `while`, `}`, or an operator, found `,`
  --> compiler/rustc_hir/src/hir.rs:69:61
   |
25 | ...                   name='_, IsAnonInPath::No
   |                              ^ expected one of 9 possible tokens

error: extern blocks must be unsafe
 --> compiler/rustc_hir/src/hir.rs:49:1
  |
5 | / extern "C" {
6 | |     fn f1(s: S);             // LifetimeName::Param, name='_, IsAnonInPath::Yes
7 | |     fn f2(s: S<'_>);         // LifetimeName::Param, name='_, IsAnonInPath::No
8 | |     fn f3<'a>(s: S<'a>);     // LifetimeName::Param, name='a, IsAnonInPath::No
9 | | }
  | |_^

warning: hidden lifetime parameters in types are deprecated
 --> compiler/rustc_hir/src/hir.rs:50:14
  |
6 |     fn f1(s: S);             // LifetimeName::Param, name='_, IsAnonInPath::Yes
  |              ^ expected lifetime parameter
  |
note: the lint level is defined here
 --> <crate attribute>:1:6
  |
1 | warn(rust_2018_idioms)
  |      ^^^^^^^^^^^^^^^^
  = note: `#[warn(elided_lifetimes_in_paths)]` implied by `#[warn(rust_2018_idioms)]`
help: indicate the anonymous lifetime
  |
6 |     fn f1(s: S<'_>);             // LifetimeName::Param, name='_, IsAnonInPath::Yes
  |               ++++

warning: hidden lifetime parameters in types are deprecated
  --> compiler/rustc_hir/src/hir.rs:62:10
   |
18 | const A: Name = Name("a");   // LifetimeName::Static, name='_, IsAnonInPath::Yes
   |          ^^^^ expected lifetime parameter
   |
help: indicate the anonymous lifetime
   |
18 | const A: Name<'_> = Name("a");   // LifetimeName::Static, name='_, IsAnonInPath::Yes
   |              ++++

error: aborting due to 4 previous errors; 2 warnings emitted

Couldn't compile the test.

failures:
    compiler/rustc_hir/src/hir.rs - hir::Lifetime (line 46)

test result: FAILED. 5 passed; 1 failed; 6 ignored; 0 measured; 0 filtered out; finished in 155.19ms

error: doctest failed, to rerun pass `-p rustc_hir --doc`
Build completed unsuccessfully in 0:39:04
  local time: Thu Mar 27 05:47:35 UTC 2025
  network time: Thu, 27 Mar 2025 05:47:35 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lcnr commented on compiler/rustc_trait_selection/src/errors.rs:
```diff
@@ -567,10 +567,14 @@ impl Subdiagnostic for AddLifetimeParamsSuggestion<'_> {
-                        if ident.name == kw::Empty && ident.span.is_empty() {
+                    let make_suggestion = |lifetime: &hir::Lifetime| {
+                        if lifetime.is_anon_in_path == IsAnonInPath::Yes
+                            && lifetime.ident.span.is_empty()
```
when is this part of the condition ever false. I would expect all `AnonInPath`  to have empty spans

nnethercote replied:
`IsAnonInPath::Yes` lifetimes are inserted in `maybe_insert_elided_lifetimes_in_path`. In that function `elided_lifetime_span` can definitely be non-empty.

However, none of those non-empty spans reach here, at least when running the UI tests. (I checked by adding an assertion.) I don't know if that's guaranteed, or if we just have insufficient test coverage. So I'd prefer to keep the test here for now; it is retained from the original code.

lcnr commented on compiler/rustc_trait_selection/src/errors.rs:
```diff
@@ -567,10 +567,14 @@ impl Subdiagnostic for AddLifetimeParamsSuggestion<'_> {
+                        {
                             format!("{}, ", self.suggestion_param_name)
-                        } else if ident.name == kw::UnderscoreLifetime && ident.span.is_empty() {
+                        } else if lifetime.ident.name == kw::UnderscoreLifetime
+                            && lifetime.ident.span.is_empty()
```
this is where changing `ident` to a source enum would be helpful

lcnr reviewed (commented):
a suggestion and one final nit

r=me after fixing the nit if u don't apply the suggestion

nnethercote commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -35,20 +35,60 @@ use crate::def_id::{DefId, LocalDefIdMap};
+/// }
+///
+/// struct Name<'a>(&'a str);    // LifetimeName::Param,  name='a, IsAnonInPath::No
+/// const A: Name = Name("a");   // LifetimeName::Static, name='_, IsAnonInPath::Yes
+/// const B: &str = "";          // LifetimeName::Static, name='_, IsAnonInPath::No
```
```rust
   // Used for fully elided lifetimes, i.e. all cases of `IsAnonInPath::Yes`
   // as well as `&Type` maybe. Idk whether these should be treated differently.
   FullyElided { insert_span: Span },
```
The "as well as `&Type` maybe" isn't possible. The `IsAnonInPath` has to be distinct otherwise `LifetimeSuggestion` won't work. You could do this:
```rust
pub enum LifetimeIdent {
    Normal { ident: Ident },
    IsAnonInPath { span: Span },
}
```
It would avoid the invalid combination of `IsAnonInPath::Yes` with a non-underscore lifetime ident, but that's the only benefit, and it adds a bunch of `match`es, etc. I don't think it's worth it.

There might still be a nicer formulation of this code, but I think this PR is worth landing in its current state.

nnethercote commented:
I didn't change the `is_empty` check for the reason explained above.

@bors r=lcnr

bors commented:
:pushpin: Commit 8d2c63f514a74752a3186b0bf3e8f3aef57dc12e has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 8d2c63f514a74752a3186b0bf3e8f3aef57dc12e -->
<!-- homu: {"type":"Approved","sha":"8d2c63f514a74752a3186b0bf3e8f3aef57dc12e","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8d2c63f514a74752a3186b0bf3e8f3aef57dc12e with merge 3f690c2257b7080cd3a8cce64e082fc972148990...
<!-- homu: {"type":"BuildStarted","head_sha":"8d2c63f514a74752a3186b0bf3e8f3aef57dc12e","merge_sha":"3f690c2257b7080cd3a8cce64e082fc972148990"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14121261862/job/39567212128)
Approved by: lcnr
Pushing 3f690c2257b7080cd3a8cce64e082fc972148990 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14121261862/job/39567212128"},"merge_sha":"3f690c2257b7080cd3a8cce64e082fc972148990"} -->

rust-timer commented:
Finished benchmarking commit (3f690c2257b7080cd3a8cce64e082fc972148990): [comparison URL](https://perf.rust-lang.org/compare.html?start=7586a9f99ad75287f9eef56c32c8508345ce0a73&end=3f690c2257b7080cd3a8cce64e082fc972148990&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=7586a9f99ad75287f9eef56c32c8508345ce0a73&end=3f690c2257b7080cd3a8cce64e082fc972148990&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7586a9f99ad75287f9eef56c32c8508345ce0a73&end=3f690c2257b7080cd3a8cce64e082fc972148990&stat=max-rss)
<details>
<summary>Results (secondary -2.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7586a9f99ad75287f9eef56c32c8508345ce0a73&end=3f690c2257b7080cd3a8cce64e082fc972148990&stat=cycles:u)
<details>
<summary>Results (secondary -0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=7586a9f99ad75287f9eef56c32c8508345ce0a73&end=3f690c2257b7080cd3a8cce64e082fc972148990&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 778.708s -> 777.401s (-0.17%)
**Artifact size**: 365.91 MiB -> 365.92 MiB (0.00%)

shepmaster commented:
Kindly have a look at https://github.com/rust-lang/rust/pull/138677 and let me know if you think I’ll have to undo these changes.

nnethercote commented:
@shepmaster: Oof, that's unfortunate timing, me removing the previously-unused distinction between `kw::Empty` and `kw::UnderscoreLifetime` just as you're adding code that depends on it. It looks like you've been working on the lint for a while, and I don't want my cleanup to block it, so we can revert it if necessary. A few questions before doing that.

- I'm confused by the diffs in #138677. In particular, `Lifetime::suggestion` looks very different to what I expect. Also `Lifetime::suggestion_position` and `LifetimeSuggestionPosition` are present, but neither of them exist on trunk. Where do they come from?

- Is there a reason the lint is done on HIR and not the AST? I would have thought the AST would be more appropriate, being closer to the original source code. Indeed, that's the reason I figured my cleanup would be fine, because HIR is further away from source and the `&u8` vs `&'_ u8` distinction didn't matter.

- I have a follow-up in #139046, currently in draft form, though it's pretty close to what would have been the final version and the commit messages explain what's going on. It eliminates the `make_suggestion` closure in much the same way your code does.

shepmaster commented:
> we can revert it if necessary

I don't believe that will be required. I've rebased on top of your changes [and pushed on through](https://github.com/rust-lang/rust/compare/master...shepmaster:rust:rework). That is my entire set of change, which will be spread out over at least three PRs (two of which are open now) — only the first 4 commits will exist in PR #138677. Specifically related to these changes, I've [continued modifying `Lifetime`](https://github.com/rust-lang/rust/commit/9b312a21b78123dd9cb0e503611bfbb97f873bfc) by transforming your new field into two. One of the new fields is (I think!) close to the [suggestion to have a source enum](https://github.com/rust-lang/rust/pull/138965#discussion_r2016307719).

> I'm confused by the diffs in [Add a new `mismatched-lifetime-syntaxes` lint #138677](https://github.com/rust-lang/rust/pull/138677). In particular, `Lifetime::suggestion` looks very different to what I expect. 

I'm not totally sure what you'd expect, but hopefully the reworked version of that function reads more cleanly?

> `Lifetime::suggestion_position` and `LifetimeSuggestionPosition` are present, but neither of them exist on trunk. Where do they come from?

They were [deleted in this very PR](https://github.com/rust-lang/rust/pull/138965/files#diff-67c494d6d292cbe0a96d420fa02cf4589c7fad5a8d9abbdfa317d49a8bdce6daL142-L152) 😉 

> Is there a reason the lint is done on HIR and not the AST? 

When I asked for help on Zulip, it [was suggested that a HIR lint would be nicer](https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp/topic/.E2.9C.94.20Adding.20new.20visitors.20vs.20piggybacking.20on.20existing/near/503645113). Based on that feedback, I scrapped my original lint changes which lived in rustc_resolve and rewrote them (extending them based on lang-team feedback) as acting on HIR.

> I have a follow-up in [Improve `Lifetime::suggestion` #139046](https://github.com/rust-lang/rust/pull/139046), currently in draft form, though it's pretty close to what would have been the final version and the commit messages explain what's going on. 

I think there's some strong similarity to my changes there. Notably, I see you are tracking if there are angle brackets, which is also something I needed to add.

> It eliminates the `make_suggestion` closure in much the same way your code does.

My [reworked code](https://github.com/rust-lang/rust/commit/9b312a21b78123dd9cb0e503611bfbb97f873bfc) continues to eliminate the the closure as well. I'm quite happy that duplicated code is being removed.

nnethercote commented:
> I've rebased on top of your changes and pushed on through

Oh, great news. I will stop working on #139046 until you are finished with the lint, though feel free to steal any ideas from it that work for you :)
 
> > `Lifetime::suggestion_position` and `LifetimeSuggestionPosition` are present, but neither of them exist on trunk. Where do they come from?
> 
> They were [deleted in this very PR](https://github.com/rust-lang/rust/pull/138965/files#diff-67c494d6d292cbe0a96d420fa02cf4589c7fad5a8d9abbdfa317d49a8bdce6daL142-L152) 😉

lol, I totally forgot I had done that. Sorry for the confusion.
